tools/ARIAtools/mask_util.py

```diff
@@ -0,0 +1,306 @@
+    """ Crop path_raster (or dataset opened with gdal) to a polygon on disk """
+    assert op.exists(path_raster) or path_raster.startswith('/vsi'), 'Invalid path to NLCD mask'
+    path_raster = gdal.BuildVRT('', path_raster)
+    ds          = gdal.Warp(path_write, path_raster, format='VRT', cutlineDSName=path_poly,
+                                             cropToCutline=True, dstNodata=9999)
+
+    if path_write: ds_vrt = gdal.BuildVRT('{}.vrt'.format(path_write), ds)
+    return ds
+
+def make_mask(ds_crop, lc):
+    # values outside crop; 0 is really just extra check
+    lc.extend([0,255])
+
+    arr = ds_crop.ReadAsArray()
+    for lclass in lc:
+        arr = np.where(arr == lclass, np.nan, arr)
+
+    arr = np.where(np.isnan(arr), np.nan, 1)
+    ds  = arr2ds(ds_crop, arr)
+    return ds
+
+def resamp(src, proj, bounds, arrshape, view=False):
+    """ Resample a dataset from src dimensions using outputs from merged_productbbox """
+    path = src.GetDescription()
+    path = path if op.exists(path) else '' # op.join(os.getcwd(), 'temp')
+    if isinstance(src, str) and op.exists(src):
+        src = gdal.Open(src, gdal.GA_ReadOnly)
+
+    ## compute geotranform
+    height, width = arrshape
+    pix_width = (bounds[2]-bounds[0]) / width
+    pix_height = (bounds[1] - bounds[3]) / height
+    gt = [bounds[0], pix_width, 0, bounds[3], 0, pix_height]
+
+    if path:
+        dst = gdal.GetDriverByName('ENVI').Create(path, width, height, 1, gdalconst.GDT_Float32)
+    else:
+        dst = gdal.GetDriverByName('MEM').Create('', width, height, 1, gdalconst.GDT_Int16)
+
+    print ('Created new')
+    dst.SetGeoTransform(gt)
+    dst.SetProjection(proj)
+    gdal.ReprojectImage(src, dst, src.GetProjection(), proj, gdalconst.GRA_NearestNeighbour)
+    print ('Reprojected')
+    del src
+    return dst
+
+def arr2ds(ds_orig, arr, noData=9999):
+    """
+    Create a new dataset identical to ds_orig except with values of 'arr'
+    Performed in memory
+    """
+    ds_orig = gdal.Open(ds_orig, gdal.GA_ReadOnly) if isinstance(ds_orig, str) else ds_orig
+    arr     = np.where(np.isnan(arr), noData, arr)
+    print ('in arr2ds')
+
+
+    if arr.ndim == 3:
+        ds_out = gdal.GetDriverByName('MEM').Create('', ds_orig.RasterXSize,
+                        ds_orig.RasterYSize, arr.shape[0], gdal.GDT_Float32)
+
+        for i in range(arr.shape[0]):
+            ds_out.GetRasterBand(i+1).WriteArray(arr[i, :, :])
+            ds_out.GetRasterBand(i+1).SetNoDataValue(noData)
+    else:
+        # ds_out = gdal.GetDriverByName('MEM').Create('', ds_orig.RasterXSize,
+        #             ds_orig.RasterYSize, 1, gdal.GDT_Float32)
+        dst = os.path.join(os.getcwd(), 'tmp')
+        print ('here')
+        ds_out = gdal.GetDriverByName('ENVI').Create(dst, ds_orig.RasterXSize,
+                    ds_orig.RasterYSize, 1, gdal.GDT_Float32)
+        print ('created dataset')
+        ds_out.GetRasterBand(1).WriteArray(arr)
```

Comment:
@bbuzz31 if i understand you mentioned offline that the problem of the exit code is happening here?
Any thoughts @sssangha?

Reply:
yup; line 196. I think I will have to edit most of the workflow to only use VRTs

Reply:
@bbuzz31 one easy way to debug the code would be to use 'tracemalloc' (see option 2 here: https://medium.com/survata-engineering-blog/monitoring-memory-usage-of-a-running-python-program-49f027e3d1ba). Just start it before line 196, and perhaps implement a try/except block such that when it runs out of memory, you will print out the traced memory.
E.g. something like this:
```import tracemalloc

tracemalloc.start()
try:
    ds_out.GetRasterBand(1).WriteArray(arr)
except:
    current, peak = tracemalloc.get_traced_memory()
    print(f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")
    tracemalloc.stop()
    raise Exception('ran out of memory')```